include/pybind11/numpy.h

```diff
@@ -610,10 +679,31 @@ class dtype : public object {
     }
 
     /// Size of the data type in bytes.
+#ifdef PYBIND11_NUMPY_1_ONLY
     ssize_t itemsize() const { return detail::array_descriptor_proxy(m_ptr)->elsize; }
+#else
+    ssize_t itemsize() const {
+        if (detail::npy_api::get().PyArray_RUNTIME_VERSION_ < 0x12) {
+            return detail::array_descriptor1_proxy(m_ptr)->elsize;
+        }
+        return detail::array_descriptor2_proxy(m_ptr)->elsize;
+    }
+#endif
 
     /// Returns true for structured data types.
+#ifdef PYBIND11_NUMPY_1_ONLY
     bool has_fields() const { return detail::array_descriptor_proxy(m_ptr)->names != nullptr; }
+#else
+    bool has_fields() const {
+        if (detail::npy_api::get().PyArray_RUNTIME_VERSION_ < 0x12) {
+            return detail::array_descriptor1_proxy(m_ptr)->names != nullptr;
+        }
+        if (num() < 0 || num() >= 2056) {
```

Comment:
Nit, should we cache num()? Not sure how expensive the call is.

Reply:
It's effectively a struct lookup.  I had assumed the compiler should be able to optimize it away (but honestly, I am not sure).

Reply:
@seberg How about this (tested locally)?

(I believe any modern compiler will optimize this.)

```diff
diff --git a/include/pybind11/numpy.h b/include/pybind11/numpy.h
index 17267800..4113d645 100644
--- a/include/pybind11/numpy.h
+++ b/include/pybind11/numpy.h
@@ -698,10 +698,11 @@ public:
         if (detail::npy_api::get().PyArray_RUNTIME_VERSION_ < 0x12) {
             return detail::array_descriptor1_proxy(m_ptr)->names != nullptr;
         }
-        if (num() < 0 || num() >= 2056) {
+        const auto *proxy = detail::array_descriptor2_proxy(m_ptr);
+        if (proxy->type_num < 0 || proxy->type_num > 2056) {
             return false;
         }
-        return detail::array_descriptor2_proxy(m_ptr)->names != nullptr;
+        return proxy->names != nullptr;
     }
 #endif
```

Reply:
Can't we just `auto _num = num()`, then use `_num`?

Reply:
Ahh, you are potentially saving one extra lookup. Okay, seems reasonable.

Reply:
What's the difference between `detail::array_descriptor2_proxy(m_ptr)` and `detail::array_descriptor_proxy(m_ptr)`? Those are inconsistent in the rewrite above.

Reply:
Ahh, okay, extra fields. Seems fine then.